src/core/shaderlib/project/project-functions.js

```diff
@@ -0,0 +1,91 @@
+/* Projection utils */
+import {COORDINATE_SYSTEM} from '../../lib/constants';
+import vec4_transformMat4 from 'gl-vec4/transformMat4';
+import vec3_sub from 'gl-vec3/subtract';
+
+function lngLatZToWorldPosition(lngLatZ, viewport) {
+  const [X, Y] = viewport.projectFlat(lngLatZ);
+  const Z = (lngLatZ[2] || 0) * viewport.scale;
+  return [X, Y, Z];
+}
+
+/**
+ * Equivalent to project_position in project.glsl
+ * projects a user supplied position to world position in the target coordinates system
+ * @param {array} position - [x, y, z]
+ * @param {object} params
+ * @param {Viewport} params.viewport - the current viewport
+ * @param {number} params.coordinateSystem - the coordinate system to project into
+ * @param {array} params.coordinateOrigin - the coordinate origin to project into
+ * @param {Matrix4} [params.modelMatrix] - the model matrix of the supplied position
+ * @param {number} [params.fromCoordinateSystem] - the coordinate system that the
+ *   supplied position is in. Default to the same as `coordinateSystem`.
+ * @param {array} [params.fromCoordinateOrigin] - the coordinate origin that the
+ *   supplied position is in. Default to the same as `coordinateOrigin`.
+ */
+export function projectPosition(
```

Comment:
Should this be added to Viewport at some point? We do not have a good project function that handles multiple coordinate systems. 

Looks like this function would benefit from a couple of quick tests?